### DISCUSSION TITLE
Request: Local Audio-In as a provider

### DISCUSSION THREAD

**Discussion by volkuhl** [music-assistant/support#1156] [Feature requests and ideas]
I don't know if that's possible, but I miss the local audio-in as a provider. It would be nice if I could select my local audio-in of the sound card (Hifiberry) as a provider.
So far, a huge thank you to all developers involved - good job!

**Comment by OzGav** [comment]
You should add this to the discussion in the music provider poll 2 https://github.com/orgs/music-assistant/discussions/1169

**Comment by bbrendon** [comment]
It's in the poll! Sounds interesting. Curious what the use cases people have in mind for this?

**Reply by volkuhl** [reply]
Audio-sources like digital-radio, turntable, tape-machine, mix-console or any other "old-school"-devices via audio-in (hifiberry) provided to any player is in my mind.

**Comment by ElektronikCZ** [comment]
I voted for this in the poll, then i realised its audio in, not out. Whats the use case for audio in?

**Reply by OzGav** [reply]
Audio out is already available if HA is on the host via the slimproto integration.

**Reply by ElektronikCZ** [reply]
There used to be a native hass local audio option

**Reply by ntninja** [reply]
Using the slimproto add-on is a nightmare of bad latency or audio artifacts (choose one or both), getting it to work requires passing a masters class… It’d awful and unreliable. I can’t imagine a native connecition from Music Assistant to the HA PluseAudio server even having half the problems (by count of potentially breaking components alone).
When initially installing Music Assistant I was super confused that the local speaker didn’t show up, to the point of mistaking “Local Device” (that appears to be called “Web” now) for the HomeAssistant Green hardware.

**Reply by nickrout** [reply]
If you have a problem with it post properly what went wrong logs etc. this is a technical forum not a bitch fest.

**Reply by ntninja** [reply]
There are no helpful logs, the problem with almost certainly with SqueezeLite. They have open issues for these things and no-one is fixing them. This only concerns MusicAssistant since ”use SqueezeLite” is the official recommendation, my “bitching” is regarding the fact that this *solution* is an extremely poor and brittle one in practice. Maybe the official communication should be “use SqueezeLite and good luck getting it to work with HomeAssistant PulseAudio, you’re gonna need it and still hate it afterwards” – that would properly communicate the status quo. (And no, since that’s obvious going to be the next short-circuit reaction, I never accused anyone here of not fixing it, I’m only complaining that this is communicated as if there is already a good solution when that plain false.)

**Reply by nickrout** [reply]
Squeezelite creates logs. Also, is you want the docs changed, submit a PR.

**Reply by ntninja** [reply]
> Squeezelite creates logs.

No, it does *not* write any logs while its hanging…

> Also, is you want the docs changed, submit a PR.

Sure, we’ll say if they accept *this*: https://github.com/music-assistant/music-assistant.io/pull/440

**Reply by nickrout** [reply]
Not up to me, but good on you for contributing

**Reply by ntninja** [reply]
lol

**Reply by Torrax** [reply]
> lol

No one's forcing you to use this btw. Its works great if youre competent with computers. I litterally help people set up home automation systems with this as a business...

**Reply by OzGav** [reply]
@ntninja you are borderline being disruptive. Your experience has not been verified as that experienced by the majority. That PR is passive aggressive as is your second last post. Contribute positively and constructively or not at all.

**Comment by nickrout** [comment]
For example a turntable. There are ways and means, but simple would be good.

**Comment by DutchJaFO** [comment]
The sonos port also has an audio input. 
I don't have a use for it myself, but it would be nice if support fire this feature was available.

The challenge is that there is no meta data, so it won't be searchable in the library.

**Comment by CaseyRo** [comment]
I'm using PCP (picoreplayer) that sets up a stream that I can listen to from music assistant. It's not "simple", but it works, and works everywhere.

**Comment by dunxd** [comment]
If this was added might it be a step towards enabling MA as a Chromecast and Airplay receiver to enable sending audio streams to MA from apps that MA otherwise couldn't access (e.g. BBC Sounds)?

**Comment by Hedda** [comment]
Would a AUX input port from an ADC (Analog to Digital Converter) Audio Capture Device also be counted as a source for "Local Audio-In"?

Personally I would really love it if was able to get audio input from such analog audio sources like a turntable (LP vinyl record player) into Music Assistant by being able to simply select "Local Audio-In" as a music provider. As per my related feature request here:

* https://github.com/orgs/music-assistant/discussions/2343

That would made it much easier and way more user-friendly function than the hoops you have to jump through to get a solution for it today. Ex:

* https://github.com/quebulm/Raspberry-Pi-Vinyl-Streamer

* https://github.com/gieljnssns/darkice-libaacplus-rpi-guide/

**Comment by Hedda** [comment]
Oh, and remember to vote for this in the new Music Provider Feature Request Poll:

* https://github.com/orgs/music-assistant/discussions/2356

**Comment by Hedda** [comment]
FYI, check out these indirectly related patches with a new feature which will add a media proxy to Home Assistant core and extend the media player info in ESPHome with a list of supported audio formats which can then be used in Home Assistant to swap out media URLs going to the ESPHome device with proxy URLs that contain the converted audio. That should allow you to at least capture the audio at best possible sample rate and then have Home Assistant automatically convert it to the higest format that the player is capable of:

* https://github.com/home-assistant/core/pull/123254
* https://github.com/esphome/aioesphomeapi/pull/925
* https://github.com/esphome/esphome/pull/7318

I wish support for this new media proxy in Home Assistant  core will implemented in Music Assistant and that someone makes a plugin or plug-and-play audio provider for getting audio stream directly from a local (USB) audio capture device into Music Assistant which can automatically use this to resample, transcode and mux it to best possible supported format before sending on to the player provider.

Again, also suggest that please check out and join this related feature request discussion:

* https://github.com/orgs/music-assistant/discussions/2343

Perhaps more active discussions in those will draw the attention and interest of a devleloper with the skills and time to make those ideas happen.

**Comment by Hedda** [comment]
@@Torrax Is the idea that music provider for "Local Audio-In" will only be able to access audio-input locally from different "sound cards" directly from Music Assistant server stand-alone and [Music Assistant add-on](https://github.com/music-assistant/home-assistant-addon) running in [Home Assistant Operating System](https://github.com/home-assistant/operating-system) as an ADC-input appliance?

* https://github.com/music-assistant/server/pull/2356

  * https://github.com/music-assistant/server/commit/33f74ef41b6883130ad854f365b0ae3c1118d2db

    _What it does_

    * _**Bring outside audio into Music Assistant.** Use your computer as the bridge._  
    * _**Any ALSA input.** Line-in jacks, USB audio interfaces, or built-in mic/line sources._  
    * _**Simple to set up.** Choose the input, start the stream, and it shows up as a source in your players._  
    * _**Custom source personalization.** Give it your own display name and thumbnail image for a polished, personalized feel in the Music Assistant UI._  

    _Use cases_

    *  _**Quick-Connect Bluetooth Receiver**  Plug a Quick Connect BT receiver into your PC’s line-in. Anyone can quickly pair their phone without having to confirm the connection, and their music instantly plays across your whole-house system._
    *  _**Announcements & Paging Microphone** Plug in a USB microphone and use it for announcements. Great for paging in a business, office, or house intercom setup._
    *  _**Vynyl Turntable/Player**   Connect your turntable (via phono preamp) directly to your PC’s line-in, and enjoy your vinyl collection throughout your Music Assistant ecosystem._

    _Requirements_

    *  _**Music Assistant** server._
    *  _**Linux host** with ALSA._
    *  _**A capture device** (line-in, USB interface, mic, etc)._

The specific use case I am wondering is the scenario of buying and connecting a compatible retail USB-soundcard direct to the USB-port on official [Home Assistant Green](https://www.home-assistant.io/green/) and/or [Home Assistant Yellow](https://www.home-assistant.io/yellow/) (or similar Raspberry Pi 4 / Raspberry Pi 5 HAOS installations) for LP-turntables.

Asking this since just spottet this new pull request by @sairon for HAOS that will add driver for Creative Sound Blaster X-Fi sound card series:

* https://github.com/home-assistant/operating-system/pull/4331

If so then I suggest also testing/verifying compatibility with these popular "high-fidelity" quality USB Audio Device ADC-adapters that uses the USB Audio Class 2.0 pipeline to support high-definition audio formats up to 192KHz and 32bits using a standard digital audio interface:

- [Behringer UFO202](https://www.behringer.com/product.html?modelCode=P0484) (with pre-amp)
- [Behringer UCA222](https://www.behringer.com/product.html?modelCode=0805-AAG) (without pre-amp, replaces the UCA202)
- [Behringer UCA202](https://www.behringer.com/behringer/product?modelCode=0805-AAC) (without pre-amp)
- [Behringer UM2](https://www.behringer.com/product.html?modelCode=0805-AAL) (with pre-amp)
- [IK Multimedia iRig Stream](https://www.ikmultimedia.com/products/irigstream/) (without pre-amp)
- [IK Multimedia iRig Stream Pro](https://www.dlxmusic.se/produkter/studio/ljudkort/externa/ik-irig-stream-pro) (with pre-amp)

Another option as "ADC" (Analog to Digital Converter) input instead of a USB Audio Device that should provide the same function are HiFiBerry's ANALOG INPUT products like "HIFIBERRY DAC+ ADC PRO", "HIFIBERRY DAC2 ADC PRO", or "HIFIBERRY DAC+ ADC" as input. As a bonus those should provide a very clean and almost the look of an all-in-one commercial appliance if using with a Raspberry Pi setup (and the [official HiFiBerry cases for Raspberry Pi](https://www.hifiberry.com/shop/#cases) for matching enclosure):
  - https://www.hifiberry.com/blog/need-some-input/ (blog post which describe the purpose of  HiFiBerry ADC HATs for Raspberry Pi GPIO)
    - https://www.hifiberry.com/shop/boards/hifiberry-dac2-adc-pro/
    - https://www.hifiberry.com/shop/boards/hifiberry-dac-adc-pro/
    - https://www.hifiberry.com/shop/boards/hifiberry-dac-adc/
      - https://www.hifiberry.com/docs/data-sheets/datasheet-dac-adc/
    - https://www.hifiberry.com/shop/0-development-kits/stage-devkit/

For reference, see these related guides/tips for Hi-Fi analog to digital used or Vinyl Record Player / LP Turntable to Music Assistant streaming:
* https://github.com/quebulm/Raspberry-Pi-Vinyl-Streamer
* https://github.com/gieljnssns/darkice-libaacplus-rpi-guide/blob/master/README.md
* https://maker.pro/raspberry-pi/projects/how-to-build-an-internet-radio-station-with-raspberry-pi-darkice-and-icecast

PS: Again there is a related discussion here but that is more about having a remote audio-in applicance instead of having it built-into the Music Assistant server which will limit you to a single audio input source and not allow you to place the audio-input device anywhere in your home:

- https://github.com/orgs/music-assistant/discussions/2343

**Reply by OzGav** [reply]
I personally haven’t done any development of this but looking at the Discord chat the initial prototype was using a UCA202 so I would say that is the general plan, to support USB input devices.

**Comment by Torrax** [comment]
@Hedda yes as OzGav outlined, this should cover pretty much the whole line on USB Sound Cards you provided. Your MA server just sees these devices as ALSA input devices so anything internal or external your system detects as an input devices should work with this plugin.

You just set the options to what ALSA device you want to stream the sound from then you can set a thumbnail image and name for the source so you can set up multiple USB cards if you want with different sources like a Vinyl Player, Bluetooth Reciever ect..

I have tested it an overall it does work pretty well, but a couple fixes I need to implement before full release.

**Reply by Hedda** [reply]
@Torrax Thanks! Would it maybe a good idea to implement a PipeWire wrapper or abstraction layer instead of talking directly to ALSA?

* https://en.wikipedia.org/wiki/PipeWire

I believe PipeWire can more or less also be used as a drop-in replacement for applications that normally use PulseAudio, JACK, or ALSA?

  * https://pipewire.org

**Reply by Torrax** [reply]
Right now the biggest problem is audio delay.

The setup I have is with the UCA202 USB Sound Card and a Logitech Bluetooth Reciever plugged into it.
When I was pressed the play button on my phone I was getting 6s+ of delay before the audio played, or stopped when I pressed pause on the source. A bunch of this comes from MAs internal buffering.

I modified a bunch of the core files temporarily to push my streams straight to source and was able to get that delay down to sub 1.5 sec. This last bit likely comes from the speaker buffer size it captures before playing which can be tuned per speaker.

I dont really want to add another layer for possible additional delay. ALSA is basically the closest you can go to the direct hardware. Looking into Pipewire, the only real benefit I can see it would add is loopback devices and basically building a pipe for audio to go through which isnt exactly what im aiming for in terms of complexity. 

Let me know if there's something I didn't catch tho and I may look at implementation down the line when things are polished up.

**Reply by Hedda** [reply]
I have not verified but thought that PipeWire was also able to do low-latency as long as use high audio sample rate?

* https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Performance#latency

As I understand PipeWire (and WirePlumber) depends on ALSA and them working in userspace making them useful.

* https://www.youtube.com/watch?v=HOsp-JlIkkY

Anyway, was more just wondering if adding PipeWire support could help achieve a better plug-and-play experience?

* https://bootlin.com/blog/an-introduction-to-pipewire/

I have not tried it on [Home Assistant Operating System](https://github.com/home-assistant/operating-system) but believe it already supports PipeWire (and PulseAudio?)?

* https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#why-another-audio-standard-linux-already-has-13-of-them

**Comment by lowlyocean** [comment]
Created a discussion for another potential use case... leveraging MA + Sendspin as a surround sound receiver for video streaming boxes:
https://github.com/orgs/music-assistant/discussions/4732

**Comment by DanielKinsman** [comment]
I think pipewire delay would be extremely minimal, it is used for real time music production (e.g. live VST effects) and is a replacement for JACK.

Pipewire would be better than alsa in my opinion because then you could grab individual playback streams from anything running linux. For example from my gaming pc I could pipe the playback from my browser into it but not the playback from the game I am currently playing. All this while still letting you do the more standard approach of grabbing the device input audio device or bluetooth a2dp input from the actual server running music assistant. Check out helvum which is a gui patch bay for pipewire.

**Comment by Hedda** [comment]
FYI; for reference, someone has submitted a draft pull request which would add support for a new source/sender role to Sendspin-clients here:

* https://github.com/Sendspin/spec/pull/52

There is also a direcly related development discussion here about how support for audio input could be added to the Sendspin protocol spec:

* https://github.com/Sendspin/spec/issues/14

PS: Noted that it looks like development of the draft pull request for adding Local Audio Source Provider for Music Assistant server has stalled, or at least @Torrax has not yet addressed the reviews with request for changes that other developers posted there:

* https://github.com/music-assistant/server/pull/2356

**Comment by grainsoflight** [comment]
Minor aside while this is still in development: Anyone still using the Darkice-> Icecast pipeline to accomplish local in, Ive found using LiquidSoap instead of Darkice leads to much lower latency (~3s, compared to Darkice which in my experience could be as high as 15s) and has the added benefit of allowing high bitrate FLAC. This may be acceptable until local-in is complete. I used this tutorial and just adapted it for Debian. https://iotsblog.wordpress.com/

(it should be noted that my turntable is connect to a raspberry pi Virtual Here USB server in one room, being served over ethernet to a Debian VM on Unraid in another room, so if your audio source is plugged directly into your linux server your latency will likely be less than mine)

**Comment by shimon-27** [comment]
Many radio stations pop up on the Internet and are not added as fast to the library, understandibly. Local-in would allow opening a browser or an app if one exists, and pipe the audio as a source like any other media. I've found a workaround with [SWYH](https://www.streamwhatyouhear.com/) to listen to streams from [accuradio](https://www.accuradio.com/) and others until there is a local-in available

**Reply by OzGav** [reply]
We have a plugin that will stream your PC audio to MA https://www.music-assistant.io/plugins/vban-receiver/

**Comment by Domoticon** [comment]
FYI, a simmilar request, marked as duplicate, for adding a CD-player as a new or dedicated filesystem :
https://github.com/orgs/music-assistant/discussions/5528#discussioncomment-17050522

**Comment by iVolt1** [comment]
I put together a local audio in plugin with Claude's help that works well. Below are some screenshots to give some idea of the capabilities.  I can create a PR if the devs are interested.

Code - https://github.com/iVolt1/server/tree/line-in/music_assistant/providers/line_in_3
git - https://github.com/iVolt1/server.git

<img width="1192" height="501" alt="image" src="https://github.com/user-attachments/assets/9e50fd41-f43c-4c55-9bf4-d26f3b935126" />

<img width="632" height="877" alt="image" src="https://github.com/user-attachments/assets/0a6cdf05-4611-4bea-b9e1-861bda41205e" />

<img width="636" height="446" alt="image" src="https://github.com/user-attachments/assets/cc270b5b-857e-4374-9248-1b5f763fe63c" />

<img width="1457" height="87" alt="image" src="https://github.com/user-attachments/assets/48608459-d76d-4ee9-a1ac-c756dea284be" />

**Reply by OzGav** [reply]
Yes but there is also this which needs work https://github.com/music-assistant/server/pull/2356

Maybe just have a look at that for inspiration and if yours works and doesn’t have that ones problems then open away

**Reply by iVolt1** [reply]
@OzGav I have looked at that PR and it seemed stalled so I started on this one that I think checks all of the boxes but let me know if you see something in the screenshots is missing or going in the wrong direction. This provider is modeled off of the current Spotify Connect provider and incorporates #3938's intentions.  I will do a pre-commit and PR before long.  Thanks.

**Comment by Hedda** [comment]
FYI, for reference, MA developers have now added a "Sendspin Source" plugin/addon to Music Assistant server 2.10 BETA releases, see:

* https://beta.music-assistant.io/plugins/sendspin-source/

  * https://github.com/music-assistant/server/tree/dev/music_assistant/providers/sendspin_source

    * https://github.com/music-assistant/server/pull/5658

And the new Music Assistant server [2.10.0 RC 7](https://github.com/music-assistant/server/releases/tag/2.10.0rc7) release retire the local audio provider in favor of the Sendspin source plugin:

* https://github.com/music-assistant/server/pull/5965 

So now all we need need is are Sendspin clients with line-level audio AUX inpout that implement support the new Sendspin source@v1 role!
